### PR TITLE
Migrate NYSE holidays to ObservedHolidayBase

### PR DESCRIPTION
**Pull request by KJhellico**
## Proposed change

Migrate NYSE holidays to `ObservedHolidayBase`.

## Type of change

- [ ] New country/market holidays support (thank you!)
- [ ] Supported country/market holidays update (calendar discrepancy fix, localization)
- [x] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new python-holidays functionality in general)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] I've added references to all holidays information sources used in this PR
- [x] The code style looks good: `make pre-commit` command generates no changes
- [x] All tests pass locally: `make test`, `make tox` (we strongly encourage adding tests to your code)

[contributing-guidelines]: https://github.com/vacanza/python-holidays/blob/beta/CONTRIBUTING.rst
[docs]: https://github.com/vacanza/python-holidays/tree/beta/docs/source

### CODE REVIEW

File: holidays/financial/ny_stock_exchange.py [hunk 1]
```diff
@@ -43,53 +30,43 @@ class NewYorkStockExchange(HolidayBase, ChristianHolidays, InternationalHolidays
...
-        elif self._is_sunday(dt):
-            dt += td(days=+1)
-            name = f"{name} (Observed)"
-
-        self._add_holiday(name, dt)
+    def _add_observed_holiday(self, name, *args) -> None:
+        dt = date(self._year, *args)
+        is_obs, _ = self._add_observed(dt, name)
+        if not is_obs:
+            self._add_holiday(name, dt)
```
Comment (arkid15r) [inline holidays/financial/ny_stock_exchange.py:46]: It didn't seem right to have two similar `_add_observed...` method. Could you check if this suggestion covers your needs:
```suggestion
    def _add_observed(self, name, *dt) -> None:  # type: ignore[override]
        dt = dt if len(dt) > 1 else dt[0]
        is_observed, _ = super()._add_observed(dt, name)  # type: ignore[arg-type, assignment]
        if not is_observed:
            self._add_holiday(name, dt)  # type: ignore[assignment]

```

You'll need to replace all `_add_observed_holiday` cases to `_add_observed`. Please note the args order: name, then date args

File: holidays/financial/ny_stock_exchange.py [hunk 2]
```diff
@@ -43,53 +30,43 @@ class NewYorkStockExchange(HolidayBase, ChristianHolidays, InternationalHolidays
...
-
-        # https://www.nyse.com/publicdocs/nyse/regulation/nyse/NYSE_Rules.pdf
-        # As per Rule 7.2.: check if next year's NYD falls on Saturday and
-        # needs to be observed on Friday (Dec 31 of previous year).
-        dec_31 = date(year, DEC, 31)
-        if self._is_friday(dec_31):
-            self._add_observed_holiday("New Year's Day", dec_31 + td(days=+1))
+        name = "New Year's Day"
+        self._add_observed_holiday("New Year's Day", JAN, 1)
+        self._add_observed(self._next_year_new_years_day, name)
```
Comment (arkid15r) [inline holidays/financial/ny_stock_exchange.py:54]:
```suggestion
        self._add_observed(name, self._next_year_new_years_day)
```

**Review by arkid15r — COMMENTED** [review]
It seems the `ObservedHolidayBase::_add_observed` needs to be overridden here:

File: holidays/financial/ny_stock_exchange.py (hunk 1 above)
Comment (KJhellico) [reply holidays/financial/ny_stock_exchange.py]: We have `_move_holiday()` for such cases, I forgot about it. :)

Comment (arkid15r) [reply holidays/financial/ny_stock_exchange.py]: That's a lot better!

**Review by arkid15r — APPROVED** [review]

**Pull request merged**